fsi/component/kinds.go

```diff
@@ -0,0 +1,588 @@
+	}
+	return nil
+}
+
+// RemoveFrom removes the component file from the directory
+func (sc *StructureComponent) RemoveFrom(dirPath string) error {
+	if err := os.Remove(filepath.Join(dirPath, "structure.json")); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+// DropDerivedValues drops derived values from the component
+func (sc *StructureComponent) DropDerivedValues() {
+	sc.Value.DropDerivedValues()
+}
+
+// LoadAndFill loads data from the component source file and assigngs it
+func (sc *StructureComponent) LoadAndFill(ds *dataset.Dataset) error {
+	if sc.Base().IsLoaded {
+		return nil
+	}
+	if sc.Value != nil {
+		sc.Base().IsLoaded = true
+		return nil
+	}
+	fields, err := sc.Base().LoadFile()
+	if err != nil {
+		return err
+	}
+	sc.Value = &dataset.Structure{}
+	if err := fill.Struct(fields, sc.Value); err != nil {
+		return err
+	}
+	if sc.Value.Schema == nil && sc.SchemaInference != nil {
+		sc.Value.Schema, err = sc.SchemaInference(ds)
+		if err != nil {
+			return err
+		}
+	}
+	if ds != nil {
+		ds.Structure = sc.Value
+	}
+	return nil
+}
+
+// CommitComponent represents a commit component
+type CommitComponent struct {
+	BaseComponent
+	Value *dataset.Commit
+}
+
+// Compare compares to another component
+func (cc *CommitComponent) Compare(compare Component) (bool, error) {
+	other, ok := compare.(*CommitComponent)
+	if !ok {
+		return false, nil
+	}
+	if err := cc.LoadAndFill(nil); err != nil {
+		return false, err
+	}
+	if err := compare.LoadAndFill(nil); err != nil {
+		return false, err
+	}
+	return compareComponentData(cc.Value, other.Value)
+}
+
+// WriteTo writes the component as a file to the directory
+func (cc *CommitComponent) WriteTo(dirPath string) error {
+	if err := cc.LoadAndFill(nil); err != nil {
+		return err
+	}
+	if cc.Value != nil && !cc.Value.IsEmpty() {
+		return writeComponentFile(cc.Value, dirPath, "commit.json")
+	}
+	return nil
+}
+
+// RemoveFrom removes the component file from the directory
+func (cc *CommitComponent) RemoveFrom(dirPath string) error {
+	if err := os.Remove(filepath.Join(dirPath, "commit.json")); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+// DropDerivedValues drops derived values from the component
+func (cc *CommitComponent) DropDerivedValues() {
+	cc.Value.DropDerivedValues()
+}
+
+// LoadAndFill loads data from the component source file and assigngs it
+func (cc *CommitComponent) LoadAndFill(ds *dataset.Dataset) error {
+	if cc.Base().IsLoaded {
+		return nil
+	}
+	if cc.Value != nil {
+		cc.Base().IsLoaded = true
+		return nil
+	}
+	fields, err := cc.Base().LoadFile()
+	if err != nil {
+		return err
+	}
+	cc.Value = &dataset.Commit{}
+	if err := fill.Struct(fields, cc.Value); err != nil {
+		return err
+	}
+	if ds != nil {
+		ds.Commit = cc.Value
+	}
+	return nil
+}
+
+// BodyComponent represents a body component
+type BodyComponent struct {
+	BaseComponent
+	Resolver       qfs.PathResolver
```

Comment:
This question has shown up a bunch in our work on filters. Do we embed a pointer to the resolver, or require it as an argument all over the place?

I'm pretty sure it's the latter, and like `context.Context`, it's an anti pattern to embed a resolver into a struct that isn't `lib.Instance`. Again, let's keep it for now & return to this question later.